### PR TITLE
`get_class_hash` cheatcode

### LINKED ISSUE #335: Add cheatcode to fetch contract ClassHash
### To what component is your bug related to?

Forge

### Feature Request

Add a cheat code to get the current class hash of a given contract address. This is needed, in part, to test the replace class functionality

### PR DESCRIPTION
**Pull request by glihm**
Closes #335 

## Introduced changes

Add a cheatcode to get the class hash associated to a given address:

```rust
let class_hash = declare(...);
// Prepare
let contract_address = deploy(...);
assert(get_class_hash(contract_address) == class_hash, 'Wrong class hash');
```

## Checklist

- [X] Linked relevant issue
- [X] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code
- [X] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: docs/src/appendix/cheatcodes/get_class_hash.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }
+}
+```
+
+The use of this cheatcode can be illustrated as following, where the `replace_class`
+syscall will change the class associated to the contract calling it:
+
+```rust
+    use snforge_std::{declare, deploy, get_class_hash, PreparedContract};
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/get_class_hash.md:41]: No need to include the imports, those are not full examples to copy-paste

File: docs/src/appendix/cheatcodes/get_class_hash.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        let class_hash = declare('Contract1');
+
+        let prepared = PreparedContract {
+            class_hash: class_hash,
+            constructor_calldata: @ArrayTrait::new()
+        };
+
+        let contract_address = deploy(prepared).unwrap();
+
+        assert(get_class_hash(contract_address) == class_hash, 'Bad class hash');
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/get_class_hash.md:54]:
```suggestion
        assert(get_class_hash(contract_address) == class_hash, 'Incorrect class hash');
```

File: docs/src/appendix/cheatcodes/get_class_hash.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        let other_class_hash = declare('OtherContract');
+
+        IUpgradeableDispatcher { contract_address }.upgrade(other_class_hash);
+
+        assert(get_class_hash(contract_address) == other_class_hash, 'Bad class hash upgrade');
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/get_class_hash.md:60]:
```suggestion
        assert(get_class_hash(contract_address) == other_class_hash, 'Incorrect class hash after upgrade');
```

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 4]
```diff
@@ -418,4 +437,11 @@ mod test {
         let err = result.unwrap_err();
         assert_eq!(err.to_string(), "Failed to convert value = yyyy to Felt252");
     }
+
+    #[test]
+    fn contract_address_from_felt_252() {
+        let felt = Felt252::from(112_233);
+        let addr = contract_address_from_felt252(&felt).unwrap();
+        assert_eq!(stark_felt_to_felt(*addr.0.key()), felt);
+    }
```
Comment (piotmag769) [inline crates/forge/src/cheatcodes_hint_processor.rs:446]: Does it test anything meaningful?

File: crates/forge/tests/integration/get_class_hash.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    constructor_calldata: @ArrayTrait::new()
+                };
+
+                let contract_address = deploy(prepared).unwrap();
+
+                assert(get_class_hash(contract_address) == class_hash, 'Bad class hash');
+
+                let hsn_class_hash = declare('HelloStarknet');
+                IUpgradeableDispatcher { contract_address }.upgrade(hsn_class_hash);
+                assert(get_class_hash(contract_address) == hsn_class_hash, 'Bad class hash upg');
```
Comment (piotmag769) [inline crates/forge/tests/integration/get_class_hash.rs:79]: Can we try to call a function from `HelloStarknet` after upgrade?

**Review by piotmag769 — COMMENTED** [review]
Really good job!

File: crates/forge/tests/integration/get_class_hash.rs (hunk 5 above)
Comment (glihm) [reply crates/forge/tests/integration/get_class_hash.rs:79]: That's a very nice idea. :+1:

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 4 above)
Comment (glihm) [reply crates/forge/src/cheatcodes_hint_processor.rs:446]: I didn't want to neglect coverage as we manipulate bytes to go from one type to another. In this PR I didn't replace it everywhere, but if it's used in more locations, we may want to ensure that the helper is converting the value correctly.

File: docs/src/appendix/cheatcodes/get_class_hash.md (hunk 3 above)
Comment (glihm) [reply docs/src/appendix/cheatcodes/get_class_hash.md:60]: I tried this before, but it's more that 31 chars. :disappointed:

File: docs/src/appendix/cheatcodes/get_class_hash.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    #[external(v0)]
+    impl IUpgradeableImpl of super::IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, class_hash: starknet::ClassHash) {
+            starknet::replace_class_syscall(class_hash).unwrap_syscall();
+        }
+    }
+
+    // ...
+```
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/get_class_hash.md:22]: I know I'm being picky, but can we get rid of the empty lines? xD

File: docs/src/appendix/cheatcodes/get_class_hash.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        let contract_address = deploy(prepared).unwrap();
+
+        assert(get_class_hash(contract_address) == class_hash, 'Incorrect class hash');
+
+        let other_class_hash = declare('OtherContract');
+
+        IUpgradeableDispatcher { contract_address }.upgrade(other_class_hash);
+
+        assert(get_class_hash(contract_address) == other_class_hash, 'Incorrect class hash upgrade');
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/get_class_hash.md:44]:
```suggestion
    assert(get_class_hash(contract_address) == other_class_hash, 'Incorrect class hash after upgrade');
```

File: crates/forge/tests/integration/get_class_hash.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
```
Comment (piotmag769) [inline crates/forge/tests/integration/get_class_hash.rs:87]: Bad -> Incorrect

File: crates/forge/tests/integration/get_class_hash.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+                let hello_dispatcher = IHelloStarknetDispatcher { contract_address };
+                hello_dispatcher.increase_balance(42);
+                assert(hello_dispatcher.get_balance() == 42, 'Fail to get balance');
```
Comment (piotmag769) [inline crates/forge/tests/integration/get_class_hash.rs:91]:
```suggestion
                assert(hello_dispatcher.get_balance() == 42, 'Invalid balance');
```

**Review by piotmag769 — APPROVED** [review]
Thanks, awesome job!

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 10]
```diff
@@ -392,6 +405,12 @@ fn write_cheatcode_panic(buffer: &mut MemBuffer, panic_data: &[Felt252]) {
         .expect("Failed to insert error in memory");
 }
 
+fn contract_address_from_felt252(felt: &Felt252) -> Result<ContractAddress, EnhancedHintError> {
+    Ok(ContractAddress(PatriciaKey::try_from(StarkFelt::new(
+        felt.to_be_bytes(),
```
Comment (piotmag769) [inline crates/forge/src/cheatcodes_hint_processor.rs:410]: I think we can use blockifier's `felt_to_stark_felt`

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 4 above)
Comment (piotmag769) [reply crates/forge/src/cheatcodes_hint_processor.rs:446]: I mean it does not hurt to test it, but I don't think we test anything meaningful here. The only thing the tested function does is struct construction , so we are basically testing blockifier's code here

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 10 above)
Comment (glihm) [reply crates/forge/src/cheatcodes_hint_processor.rs:410]: You're totally right. Thanks!

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 4 above)
Comment (glihm) [reply crates/forge/src/cheatcodes_hint_processor.rs:446]: Sorry I didn't get you well. Test was removed due to your suggestion to convert it using blockifier `felt_to_stark_felt`. :+1:

File: docs/src/appendix/cheatcodes/get_class_hash.md (hunk 6 above)
Comment (glihm) [reply docs/src/appendix/cheatcodes/get_class_hash.md:22]: Fine to be picky, thanks a lot for the feedback.

File: docs/src/appendix/cheatcodes/get_class_hash.md (hunk 7 above)
Comment (glihm) [reply docs/src/appendix/cheatcodes/get_class_hash.md:44]: This is too long @piotmag769. Do you have any alternative?
"Incorrect upgrade class hash" can fit. Is it fine for you?

File: crates/forge/tests/integration/get_class_hash.rs (hunk 8 above)
Comment (glihm) [reply crates/forge/tests/integration/get_class_hash.rs:87]: Sorry, messed up those ones with a copy-paste.

**Review by Radinyn — APPROVED** [review]
🚀🚀

**Pull request merged**